File: Week_06-动态规划/403.青蛙过河.cpp
/*
 * @lc app=leetcode.cn id=403 lang=cpp
 *
 * [403] 青蛙过河
 */
//参考题解：https://leetcode-cn.com/problems/frog-jump/solution/cji-yi-hua-sou-suo-kuai-yu-96yi-shang-de-yong-hu-b/
// @lc code=start
class Solution {
public:
    bool canCross(vector<int>& stones) {
        unordered_set<int> h; //存放所有石头的位置
        for(auto x:stones) h.insert(x);
        unordered_map<long long, bool> memo;
        function<bool(int, int)> dfs = [&] (int x, int y) {
            if(y<=0 || !h.count(x)) return false;
            if(x==1 && y==1) return true;

            long long t = (long long)x<<32|y;
            if(memo.count(t)) return memo[t];

            if(dfs(x-y,y) || dfs(x-y, y-1) || dfs(x-y, y+1))
                return memo[t] = true;
            return memo[t] = false;
        };

        for (int i=1; i<=1001; i++)
            if (dfs(stones.back(), i))
                return true;
        return false;
    }
};
// @lc code=end

File: Week_06-动态规划/403.青蛙过河.py
#
# @lc app=leetcode.cn id=403 lang=python
#
# [403] 青蛙过河
#

# @lc code=start
from collections import defaultdict


class Solution(object):
    def canCross(self, stones):
        """
        :type stones: List[int]
        :rtype: bool
        """
        '''
        #普通dfs，超时报错
        def dfs(pos,step):
            if pos==stones[-1]: return True
            for d in [-1,0,1]:
                if step+d>0 and pos+step+d in set(stones):
                    if dfs(pos+step+d, step+d):
                        return True
            return False
        pos, step = 0, 0
        return dfs(pos, step)
        '''

        # 动态规划
        # 定义dp为字典，其中key=stone, value={可以到达的跳跃步长组成的集合},
        # 那么能够到达stone等价于dp[stone]非空。
        set_stones = set(stones)
        dp = defaultdict(set)
        dp[0] = {0}
        for s in stones:
            for step in dp[s]:
                for d in [-1, 0, 1]:
                    if step+d > 0 and s+step+d in set_stones:
                        dp[s+step+d].add(step+d)
        return len(dp[stones[-1]]) > 0

# @lc code=end

